projects/kbd-2.6.4/docs/man/man1/dumpkeys.1
.\" @(#)loadkeys.1 1.0 93/09/1 RK
.TH DUMPKEYS 1 "1 Sep 1993" "kbd"
.SH NAME
dumpkeys \- dump keyboard translation tables
.SH SYNOPSIS
.B dumpkeys
[OPTIONS]
.SH DESCRIPTION
.IX "dumpkeys command" "" "\fLdumpkeys\fR command"
.LP
.B dumpkeys
writes, to the standard output, the current contents of the keyboard
driver's translation tables, in the format specified by
.BR keymaps (5).
.LP
Using the various options, the format of the output can be controlled
and also other information from the kernel and the programs
.BR dumpkeys (1)
and
.BR loadkeys (1)
can be obtained.
.SH OPTIONS
.TP
.B \-h \-\-help
Prints the program's version number and a short usage message to the
program's standard error output and exits.
.TP
.B \-i \-\-short-info
Prints some characteristics of the kernel's keyboard driver. The items
shown are:
.LP
.RS
Keycode range supported by the kernel
.LP
.RS
This tells what values can be used after the
.B keycode
keyword in keytable files. See
.BR keymaps (5)
for more information and the syntax of these files.
.RE
.LP
Number of actions bindable to a key
.LP
.RS
This tells how many different actions a single key can output using
various modifier keys. If the value is 16 for example, you can define up
to 16 different actions to a key combined with modifiers. When the value
is 16, the kernel probably knows about four modifier keys, which you can
press in different combinations with the key to access all the bound
actions.
.RE
.LP
Ranges of action codes supported by the kernel
.LP
.RS
This item contains a list of action code ranges in hexadecimal notation.
These are the values that can be used in the right hand side of a key
definition, ie. the
.IR vv 's
in a line
.LP
.RS
.B keycode
.I xx
=
.I vv vv vv vv
.RE
.LP
(see
.BR keymaps (5)
for more information about the format of key definition lines).
.BR dumpkeys (1)
and
.BR loadkeys (1)
support a symbolic notation, which is preferable to the numeric one, as
the action codes may vary from kernel to kernel while the symbolic names
usually remain the same. However, the list of action code ranges can be
used to determine, if the kernel actually supports all the symbols
.BR loadkeys (1)
knows, or are there maybe some actions supported by the kernel that
have no symbolic name in your
.BR loadkeys (1)
program. To see this, you compare the range list with the action symbol
list, see option
.B --long-info
below.
.RE
.LP
Number of function keys supported by kernel
.LP
.RS
This tells the number of action codes that can be used to output
strings of characters. These action codes are traditionally bound to
the various function and editing keys of the keyboard and are defined
to send standard escape sequences. However, you can redefine these to
send common command lines, email addresses or whatever you like.
Especially if the number of this item is greater than the number of
function and editing keys in your keyboard, you may have some "spare"
action codes that you can bind to AltGr-letter combinations, for example,
to send some useful strings. See
.BR loadkeys (1)
for more details.
.RE
.LP
Function strings
.LP
.RS
You can see you current function key definitions with the command
.LP
.RS
.B dumpkeys --funcs-only
.RE
.LP
.RE
.RE
.LP
.TP
.B \-l \-s \-\-long-info
This option instructs
.B dumpkeys
to print a long information listing. The output is the same as with the
.B --short-info
appended with the list of action symbols supported by
.BR loadkeys (1)
and
.BR dumpkeys (1),
along with the symbols' numeric values.
.LP
.TP
.B \-n \-\-numeric
This option causes
.B dumpkeys
to by-pass the conversion of action code values to symbolic notation and
to print the in hexadecimal format instead.
.LP
.TP
.B \-f \-\-full-table
This makes
.B dumpkeys
skip all the short-hand heuristics (see
.BR keymaps (5))
and output the key bindings in the canonical form. First a keymaps
line describing the currently defined modifier combinations
is printed. Then for each key a row with a column for each
modifier combination is printed. For
example, if the current keymap in use uses seven modifiers,
every row will have seven action code columns. This format
can be useful for example to programs that post-process the
output of
.BR dumpkeys .
.LP
.TP
.BI \-S shape " " " " \-\-shape= shape
Available shapes:
.LP
.RS
.B 2
default output.
.RE
.LP
.RS
.B 4
one line for each keycode.
.RE
.LP
.RS
.B 8
one line for each (modifier,keycode) pair.
.RE
.LP
.RS
.B 16
one line for each keycode until 1st hole.
.RE
.LP
.TP
.B \-1 \-\-separate-lines
This forces
.B dumpkeys
to write one line per (modifier,keycode) pair. It prefixes the word
.I plain
for plain keycodes.
.LP
.TP
.B \-t \-\-funcs-only
When this option is given,
.B dumpkeys
prints only the function key string definitions. Normally
.B dumpkeys
prints both the key bindings and the string definitions.
.LP
.TP
.B \-k \-\-keys-only
When this option is given,
.B dumpkeys
prints only the key bindings. Normally
.B dumpkeys
prints both the key bindings and the string definitions.
.LP
.TP
.B \-d \-\-compose-only
When this option is given,
.B dumpkeys
prints only the compose key combinations.
This option is available only if your kernel has compose key support.
.LP
.TP
.BI \-c charset " " " " \-\-charset= charset
This instructs
.B dumpkeys
to interpret character code values according to the specified character
set. This affects only the translation of character code values to
symbolic names. Valid values for
.I charset
currently are
.BR iso-8859-X ,
Where X is a digit in 1-9.  If no
.I charset
is specified,
.B iso-8859-1
is used as a default.
This option produces an output line `charset "iso-8859-X"', telling
loadkeys how to interpret the keymap. (For example, "division" is
0xf7 in iso-8859-1 but 0xba in iso-8859-8.)
.LP
.TP
.BI \-C dev " " " " \-\-console= dev
The affected console device can be specified using the
.I -C
(or
.I --console
) option. This option supports exactly one device name.
.LP
.TP
.B \-v \-\-verbose
Turn on verbose output.
.LP
.TP
.B \-V \-\-version
Prints version number and exits.
.LP
.SH FILES
.TP
.I @DATADIR@/keymaps
The recommended directory for keytable files.
.LP
.SH "SEE ALSO"
.BR loadkeys (1),
.BR keymaps (5)



projects/kbd-2.6.4/src/dumpkeys.c
/*
 * dumpkeys.c
 *
 * derived from version 0.81 - aeb@cwi.nl
 */
#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <fcntl.h>
#include <getopt.h>
#include <unistd.h>
#include <linux/kd.h>
#include <linux/keyboard.h>
#include <sys/ioctl.h>
#include <string.h>
#include <errno.h>
#include <sysexits.h>
#include "ksyms.h"
#include "modifiers.h"

#include "libcommon.h"

static int fd;

static void __attribute__((noreturn))
usage(int rc, const struct kbd_help *options)
{
	const struct kbd_help *h;

	fprintf(stderr, _("Usage: %s [option...]\n"), get_progname());

	if (options) {
		int max = 0;

		fprintf(stderr, "\n");
		fprintf(stderr, _("Options:"));
		fprintf(stderr, "\n");

		for (h = options; h && h->opts; h++) {
			int len = (int) strlen(h->opts);
			if (max < len)
				max = len;
		}
		max += 2;

		for (h = options; h && h->opts; h++)
			fprintf(stderr, "  %-*s %s\n", max, h->opts, h->desc);
	}

	fprintf(stderr, "\n");
	fprintf(stderr, _("Available charsets: "));
	lk_list_charsets(stderr);
	fprintf(stderr, "\n");

	fprintf(stderr, _("Available shapes:\n"
	                  "  2  - default output;\n"
	                  "  4  - one line for each keycode;\n"
	                  "  8  - one line for each (modifier,keycode) pair;\n"
	                  "  16 - one line for each keycode until 1st hole.\n"
	                 ));
	fprintf(stderr, "\n");
	fprintf(stderr, _("Report bugs to authors.\n"));
	fprintf(stderr, "\n");

	exit(rc);
}

int main(int argc, char *argv[])
{
	int c, rc;
	int kbd_mode;

	char long_info       = 0;
	char short_info      = 0;
	char numeric         = 0;
	lk_table_shape table = LK_SHAPE_DEFAULT;
	char funcs_only      = 0;
	char keys_only       = 0;
	char diac_only       = 0;
	char *console        = NULL;

	struct lk_ctx *ctx;

	set_progname(argv[0]);
	setuplocale();

	const char *short_opts = "hilvsnf1tkdS:c:C:V";
	const struct option long_opts[] = {
		{ "help", no_argument, NULL, 'h' },
		{ "short-info", no_argument, NULL, 'i' },
		{ "long-info", no_argument, NULL, 'l' },
		{ "numeric", no_argument, NULL, 'n' },
		{ "full-table", no_argument, NULL, 'f' },
		{ "separate-lines", no_argument, NULL, '1' },
		{ "shape", required_argument, NULL, 'S' },
		{ "funcs-only", no_argument, NULL, 't' },
		{ "keys-only", no_argument, NULL, 'k' },
		{ "compose-only", no_argument, NULL, 'd' },
		{ "charset", required_argument, NULL, 'c' },
		{ "console", required_argument, NULL, 'C' },
		{ "verbose", no_argument, NULL, 'v' },
		{ "version", no_argument, NULL, 'V' },
		{ NULL, 0, NULL, 0 }
	};
	const struct kbd_help opthelp[] = {
		{ "-i, --short-info",       _("display information about keyboard driver.") },
		{ "-l, -s, --long-info",    _("display above and symbols known to loadkeys.") },
		{ "-n, --numeric",          _("display keytable in hexadecimal notation.") },
		{ "-f, --full-table",       _("don't use short-hand notations, one row per keycode.") },
		{ "-1, --separate-lines",   _("one line per (modifier,keycode) pair.") },
		{ "-S, --shape={2|4|8|16}", "" },
		{ "-t, --funcs-only",       _("display only the function key strings.") },
		{ "-k, --keys-only",        _("display only key bindings.") },
		{ "-d, --compose-only",     _("display only compose key combinations.") },
		{ "-c, --charset=CHARSET",  _("interpret character action codes to be from the specified character set.") },
		{ "-C, --console=DEV",      _("the console device to be used.") },
		{ "-v, --verbose",          _("be more verbose.") },
		{ "-V, --version",          _("print version number.")     },
		{ "-h, --help",             _("print this usage message.") },
		{ NULL, NULL }
	};

	ctx = lk_init();
	if (!ctx) {
		exit(EXIT_FAILURE);
	}

	while ((c = getopt_long(argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch (c) {
			case 'i':
				short_info = 1;
				break;
			case 's':
			case 'l':
				long_info = 1;
				break;
			case 'n':
				numeric = 1;
				break;
			case 'f':
				table = LK_SHAPE_FULL_TABLE;
				break;
			case '1':
				table = LK_SHAPE_SEPARATE_LINES;
				break;
			case 'S':
				table = atoi(optarg);
				break;
			case 't':
				funcs_only = 1;
				break;
			case 'k':
				keys_only = 1;
				break;
			case 'd':
				diac_only = 1;
				break;
			case 'v':
				lk_set_log_priority(ctx, LOG_INFO);
				break;
			case 'c':
				if ((lk_set_charset(ctx, optarg)) != 0) {
					fprintf(stderr, _("unknown charset %s - ignoring charset request\n"),
					        optarg);
					usage(EX_USAGE, opthelp);
				}
				printf("charset \"%s\"\n", optarg);
				break;
			case 'C':
				console = optarg;
				break;
			case 'V':
				print_version_and_exit();
				break;
			case 'h':
				usage(EXIT_SUCCESS, opthelp);
				break;
			case '?':
				usage(EX_USAGE, opthelp);
				break;
		}
	}

	if (optind < argc)
		usage(EX_USAGE, opthelp);

	if ((fd = getfd(console)) < 0)
		kbd_error(EXIT_FAILURE, 0, _("Couldn't get a file descriptor referring to the console."));

	/* check whether the keyboard is in Unicode mode */
	if (ioctl(fd, KDGKBMODE, &kbd_mode))
		kbd_error(EXIT_FAILURE, errno, _("Unable to read keyboard mode"));

	if (kbd_mode == K_UNICODE) {
		lk_set_parser_flags(ctx, LK_FLAG_PREFER_UNICODE);
	}

	if ((rc = lk_kernel_keymap(ctx, fd)) < 0)
		goto fail;

	if (short_info || long_info) {
		lk_dump_summary(ctx, stdout, fd);

		if (long_info) {
			printf(_("Symbols recognized by %s:\n(numeric value, symbol)\n\n"),
			       get_progname());
			lk_dump_symbols(ctx, stdout);
		}
		exit(EXIT_SUCCESS);
	}

#ifdef KDGKBDIACR
	if (!diac_only) {
#endif
		if (!funcs_only) {
			lk_dump_keymap(ctx, stdout, table, numeric);
		}
#ifdef KDGKBDIACR
	}

	if (!funcs_only && !keys_only)
		lk_dump_diacs(ctx, stdout);
#endif

fail:
	lk_free(ctx);
	close(fd);

	if (rc < 0)
		exit(EXIT_FAILURE);

	exit(EXIT_SUCCESS);
}
